lib/chipmunk/package_storage.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module Chipmunk
+  # PackageStorage is a factory for proxies to a Package's underlying storage.
+  # An instance supports a named set of format types, mapping each to a storage
+  # proxy class and creating an appropriate proxy instance for a package.
+  class PackageStorage
+
+    # Create a PackageStorage instance.
+    #
+    # @param volumes [Array<Chipmunk::Volume>] the storage volumes where packages are stored
+    def initialize(volumes:)
+      @volumes = Hash[volumes.map {|volume| [volume.name, volume] }]
+    end
+
+    # Retrieve the storage proxy for a given package. The package must be stored and of a registered format.
+    # @param package [Package] the package for which to retrieve a storage proxy
+    # @raise [Chipmunk::PackageNotStoredError] if the package is not stored
+    # @raise [Chipmunk::UnsupportedFormatError] if the package format is not registered
+    def for(package)
+      raise Chipmunk::PackageNotStoredError, package unless package.stored?
+
+      volume_for(package).get(package.storage_path)
+    end
+
+    # Move the source archive into preservation storage and update the package's
+    # storage_volume and storage_path accordingly.
+    def write(package, source)
+      if package.format == Chipmunk::Bag.format
+        move_bag(package, source)
```

Comment:
This seems like it should be in `Volume`; was the thought that here makes more sense since its' coordinating between the volume and the source?

Reply:
After discussing I think this seems fine -- `Volume` is really just about the filesystem; it's used by both `PackageStorage` and `IncomingStorage`, and those are what handle the formats & layouts.

Reply:
Yes, exactly. The Volume knows nothing about Package for now, only paths and on-disk representation/proxy classes. Now, clearly, a format-specific move method is wacky, and a Volume could embody its layout as well as the proxy format, but `#write` gives us a nice enough API behind which to vary who exactly does the path calculation and how the file operations are completed.